App Insights/Operational Insights Purge API fixes

alexeldeib opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

Commit:
Purge API bugs

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-sdk-for-ruby

alexeldeib commented:
Apologies for messing with an API in the stable folder, but I went through a few rounds with swagger in the interim and this is really not accurate to our shape. I thought we could use long-running-operation and clients would do the "smart" thing, but that was a bit naive. I've made the changes I think are necessary to fix the shape.

I've also attached example (zipped) fiddler requests using the API, just as a sanity check for review purposes that this looks correct now. 
[purgestatus.zip](https://github.com/Azure/azure-rest-api-specs/files/2094956/purgestatus.zip)

annatisch commented:
Thanks @alexeldeib - could you fix the duplicate "200" response key in "examples/ComponentsPurgeStatus.json"?

alexeldeib committed:
remove duplicate 200

alexeldeib commented:
@annatisch whoops -- fixed!

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-libraries-for-java

annatisch commented:
Thanks @alexeldeib!
We still have a duplicate model reference:
```
ERROR (Fatal/DuplicateModelCollsion): Duplicated model name with non-identical definitions
    - file:///tmp/tmpssj7vnyg/rest/specification/applicationinsights/resource-manager/Microsoft.Insights/stable/2015-05-01/workbooks_API.json:258:4 ($.definitions.Resource)
    - file:///tmp/tmpssj7vnyg/rest/specification/applicationinsights/resource-manager/Microsoft.Insights/stable/2015-05-01/webTests_API.json:285:4 ($.definitions.Resource)
    - file:///tmp/tmpssj7vnyg/rest/specification/applicationinsights/resource-manager/Microsoft.Insights/stable/2015-05-01/components_API.json:321:4 ($.definitions.Resource)
```
There are two possible fixes for this. You can either:
- Make sure each definition of Resource is identical.
- Have only a single common definition of Resource that all the files reference.

alexeldeib committed:
validation fixes

alexeldeib committed:
additional autorest validation

alexeldeib commented:
Sorry, I realized I ran validation on the wrong specs >.< Went ahead and fixed a few more issues I introduced as well as some existing warnings. The remainder seem mostly harmless to me -- some of these I can definitely fix, but I would want to sync with some of the other contributors to this spec before making edits outside my domain (primarily purge + data-plane).

alexeldeib committed:
remove duplicate key again

annatisch commented:
Thanks @alexeldeib!
If the remaining warnings are unrelated to the content of this PR that's fine.
Is it possible to fix the remain error for the examples?
```
{
  "type": "Error",
  "code": "XmsExamplesRequired",
  "message": "Please provide x-ms-examples describing minimum/maximum property set for response/request payloads for operations. Operation: 'Workspaces_PurgeStatus'",
  "id": "D5001",
  "validationCategory": "Documentation",
  "providerNamespace": "Microsoft.OperationalInsights",
  "resourceType": "workspaces",
  "sources": [
    "file:///home/travis/build/Azure/azure-rest-api-specs/specification/operationalinsights/resource-manager/Microsoft.OperationalInsights/stable/2015-03-20/OperationalInsights.json:680:6 ($.paths[\"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.OperationalInsights/workspaces/{workspaceName}/operations/{purgeId}\"].get)"
```

alexeldeib committed:
Link example for Purge status

alexeldeib committed:
operation name

alexeldeib committed:
enum casing

alexeldeib commented:
Linked the example! Interestingly, i'm not seeing that error when doing my local validation? I usually do `autorest --azure-validator` and then `oav validate-spec <spec.json>; oav validate-example <spec.json>`. I thought the error you mentioned is normally picked up by `autorest --azure-validator`, but removing the example didn't change my output? Is there another validation step I'm missing?

annatisch commented on specification/operationalinsights/resource-manager/Microsoft.OperationalInsights/stable/2015-03-20/OperationalInsights.json:
```diff
@@ -661,8 +660,47 @@
               "$ref": "#/definitions/WorkspacePurgeResponse"
+            },
+            "headers": {
+              "x-ms-status-location": {
```
@alexeldeib - this looks like it's using long-running-operation mechanism but with a non-standard header name? Is this a new API or an old API that has been public for some time?

alexeldeib replied:
This is an existing public API. The swagger modeling used x-ms-long-running-operation originally because I thought we were aligned with the long-running-operation mechanism, but as you've noticed that's unfortunately not the case. I can discuss the possibility of revising this with my team (maybe for a future version?) but the API is indeed live (see the .zip Fiddler files attached to the PR if you'd like an example to compare against).

annatisch replied:
Thanks @alexeldeib! Indeed in a future version this will need to be brought in line with ARM guidelines - it would be best to consult with @ravbhatnagar.
Does that mean the other new APIs in this PR (e.g. Get status for ongoing purge operation) have also been live for some time? Or have they been added in order to support the unconventional long-running operations?

alexeldeib replied:
Purge + Purge status have both existed for a while now. Originally these were modeled as two separate requests in swagger. Here's the original [Purge status operation](https://github.com/Azure/azure-rest-api-specs/blob/57d4deeb0c569c395a295f7c312aed09ad6e30d6/specification/applicationinsights/resource-manager/Microsoft.Insights/stable/2015-05-01/components_API.json#L286).

We modified the API due to a security issue, splitting out the filter field into operator/value in #2808. Piggybacking on the breaking change I was already making, I used the chance to move us to use x-ms-long-running-operation. At the time it wasn't clear to me we didn't align to this model. 

Beyond that, the [original modeling](https://github.com/Azure/azure-rest-api-specs/blob/57d4deeb0c569c395a295f7c312aed09ad6e30d6/specification/applicationinsights/resource-manager/Microsoft.Insights/stable/2015-05-01/components_API.json#L286) with the two operations split was missing the headers, which this PR adds. Apologies for making a bit of a mess here...

annatisch commented:
Thanks @alexeldeib - the rest looks good - I'm guessing you're already aware that this PR involves breaking changes for both client SDKs?

alexeldeib commented:
Correct -- feel free to tag this as such (I wasn't sure if that's something I should do, but I noticed it's used elsewhere).
